src/cockpit/channels/filesystem.py

```diff
@@ -268,3 +287,226 @@ def do_close(self):
+        @functools.lru_cache()
+        def get_user(uid: int) -> 'str | int':
+            try:
+                return pwd.getpwuid(uid).pw_name
+            except KeyError:
+                return uid
+
+        @functools.lru_cache()
+        def get_group(gid: int) -> 'str | int':
+            try:
+                return grp.getgrgid(gid).gr_name
+            except KeyError:
+                return gid
+
+        stat_types = {stat.S_IFREG: 'reg', stat.S_IFDIR: 'dir', stat.S_IFLNK: 'lnk', stat.S_IFCHR: 'chr',
+                      stat.S_IFBLK: 'blk', stat.S_IFIFO: 'fifo', stat.S_IFSOCK: 'sock'}
+        available_stat_getters = {
+            'type': lambda buf: buf and stat_types.get(stat.S_IFMT(buf.st_mode)),
+            'tag': tag_from_stat,
+            'mode': lambda buf: stat.S_IMODE(buf.st_mode),
+            'size': lambda buf: buf.st_size,
+            'uid': lambda buf: buf.st_uid,
+            'gid': lambda buf: buf.st_gid,
+            'mtime': lambda buf: buf.st_mtime,
+            'user': lambda buf: get_user(buf.st_uid),
+            'group': lambda buf: get_group(buf.st_gid),
+        }
+        stat_getters = tuple((key, available_stat_getters.get(key, lambda _: None)) for key in attrs)
+
+        def get_attrs(fd: int, name: str) -> 'JsonDict | None':
+            try:
+                buf = os.lstat(name, dir_fd=fd) if name else os.fstat(fd)
+            except FileNotFoundError:
+                return None
+            except OSError:
+                return {name: None for name, func in stat_getters}
+
+            result = {key: func(buf) for key, func in stat_getters}
+
+            if 'target' in result and stat.S_IFMT(buf.st_mode) == stat.S_IFLNK:
+                with contextlib.suppress(OSError):
+                    result['target'] = os.readlink(name, dir_fd=fd)
+
+            return result
+
+        return get_attrs
+
+    def send_update(self, updates: JsonDict, *, reset: bool = False) -> None:
+        if reset:
+            if set(self.current_value) & set(updates):
+                # if we have an overlap, we need to do a proper reset
+                self.send_json({name: None for name in self.current_value}, partial=True)
+                self.current_value = {'partial': True}
+                updates.update(partial=None)
+            else:
+                # otherwise there's no overlap: we can just remove the old keys
+                updates.update({key: None for key in self.current_value})
+
+        json_merge_and_filter_patch(self.current_value, updates)
+        if updates:
+            self.send_json(updates)
+
+    def process_update(self, updates: 'set[str]', *, reset: bool = False) -> None:
+        assert self.fd is not None
+
+        entries: JsonDict = {name: self.getattrs(self.fd, name) for name in updates}
+        info = entries.pop('', {})
+        assert isinstance(info, dict)  # fstat() will never fail with FileNotFoundError
```

Comment:
We've just extracted `info` as a value from the `entries` dictionary we built above.  The values in that dictionary come from the result of the `self.getattrs()` call, which returns a dictionary or None.  None is only returned in the case of `FileNotFoundError`, which will never be true for the root path itself (since we call `fstat()` in that case, not `lstat()` and `fstat()` can't fail with `ENOENT`).

The logic there is all fine and good, but mypy is (understandably) unable to infer that via the types alone, so we use this assert to help with type narrowing.  Otherwise, when we treat `info` as a dict below (by assigning to it) mypy would warn that it might be `None`.

I'm not thrilled that this was necessary and I tried a few other ways to get around this, but ultimately the requirement to have this here is fairly understandable...

Reply:
ah, so this is a mypy nudge, thanks.